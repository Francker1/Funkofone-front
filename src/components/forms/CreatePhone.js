import React, { useState } from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import useForm from '../../hooks/useForm';
import NavBar from '../navbar/NavBar';
import { api } from '../../helpers/methodsPhone';

const CreatePhone = () => {
  const [ui, setUi] = useState({
    message: '',
    id: '',
  });

  const { message, id } = ui;
  // use a custom hook to set fields in form and get data
  const [formValues, handleInputChange] = useForm({
    name: '',
    price: '',
    manufacturer: '',
    color: '',
    model: '',
    detail: '',
    ram: '',
    size: '',
    screen: '',
    processor: '',
  });

  const {
    name,
    price,
    manufacturer,
    model,
    color,
    detail,
    ram,
    size,
    screen,
    processor,
  } = formValues;

  const handleSubmit = async (e) => {
    e.preventDefault();

    const resp = await api.createPhone(formValues);

    // Handle also error response and set new message
    // Set message to ui, but also is possible reset form or change disabled status to button Create
    if (resp.status === 201) {
      setUi({
        message: `Phone created!!`,
        id: resp.data._id,
      });
    }
  };

  return (
    <>
      <NavBar />
      <Container className="list-container">
        <Row>
          <Col className="mb-4">
            <h1>Create Phone</h1>
            <Link to="/phones">Back to All Phones</Link>
          </Col>
        </Row>
        <Row>
          <Col>
            <form onSubmit={handleSubmit}>
              <div className="form-row">
                <div className="form-group col-md-6">
                  <label htmlFor="phoneName">Phone name</label>
                  <input
                    type="text"
                    name="name"
                    id="phoneName"
                    className="form-control"
                    placeholder="Phone name"
                    autoComplete="off"
                    value={name || ''}
                    onChange={handleInputChange}
                    required
                  />
                </div>
                <div className="form-group col-md-6">
                  <label htmlFor="phonePrice">Price</label>
                  <input
                    type="number"
                    name="price"
                    id="phonePrice"
                    className="form-control"
                    placeholder="Phone price"
                    autoComplete="off"
                    value={price || ''}
                    onChange={handleInputChange}
                    required
                  />
                </div>
              </div>
              <div className="form-row">
                <div className="form-group col-md-4">
                  <label htmlFor="phoneManufacturer">Manufacturer</label>
                  <input
                    type="text"
                    name="manufacturer"
                    id="phoneManufacturer"
                    className="form-control"
                    placeholder="Phone manufacturer"
                    autoComplete="off"
                    value={manufacturer || ''}
                    onChange={handleInputChange}
                    required
                  />
                </div>
                <div className="form-group col-md-4">
                  <label htmlFor="phoneModel">Model</label>
                  <input
                    type="text"
                    name="model"
                    id="phoneModel"
                    className="form-control"
                    placeholder="Phone model"
                    autoComplete="off"
                    value={model || ''}
                    onChange={handleInputChange}
                    required
                  />
                </div>
                <div className="form-group col-md-4">
                  <label htmlFor="phoneColor">Color</label>
                  <input
                    type="text"
                    name="color"
                    id="phoneColor"
                    className="form-control"
                    placeholder="Phone color"
                    autoComplete="off"
                    value={color || ''}
                    onChange={handleInputChange}
                    required
                  />
                </div>
              </div>
              <div className="form-row">
                <div className="form-group col">
                  <label htmlFor="phoneDetail">Detail</label>
                  <textarea
                    className="form-control"
                    name="detail"
                    id="phoneDetail"
                    placeholder="Phone detail"
                    autoComplete="off"
                    value={detail || ''}
                    onChange={handleInputChange}
                    required
                  />
                </div>
              </div>
              <div className="form-row">
                <div className="form-group col-md-3">
                  <label htmlFor="phoneRam">RAM</label>
                  <input
                    type="text"
                    name="ram"
                    id="phoneRam"
                    className="form-control"
                    placeholder="RAM"
                    autoComplete="off"
                    value={ram || ''}
                    onChange={handleInputChange}
                  />
                </div>

                <div className="form-group col-md-3">
                  <label htmlFor="phoneSize">Size</label>
                  <input
                    type="text"
                    name="size"
                    id="phoneSize"
                    className="form-control"
                    placeholder="Phone size"
                    autoComplete="off"
                    value={size || ''}
                    onChange={handleInputChange}
                  />
                </div>

                <div className="form-group col-md-3">
                  <label htmlFor="phoneScreen">Screen</label>
                  <input
                    type="text"
                    name="screen"
                    id="phoneScreen"
                    className="form-control"
                    placeholder="Phone screen"
                    autoComplete="off"
                    value={screen || ''}
                    onChange={handleInputChange}
                  />
                </div>

                <div className="form-group col-md-3">
                  <label htmlFor="phoneProcessor">Processor</label>
                  <input
                    type="text"
                    name="processor"
                    id="phoneProcessor"
                    className="form-control"
                    placeholder="Phone processor"
                    autoComplete="off"
                    value={processor || ''}
                    onChange={handleInputChange}
                  />
                </div>
              </div>
              <button type="submit" className="btn btn-primary">
                Create
              </button>
              {message && (
                <p>
                  {message} <Link to={`/phones/${id}`}>See phone</Link>
                </p>
              )}
            </form>
          </Col>
        </Row>
      </Container>
    </>
  );
};

export default CreatePhone;
