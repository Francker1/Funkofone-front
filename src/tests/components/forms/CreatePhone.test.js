/* eslint-disable */
import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import CreatePhone from '../../../components/forms/CreatePhone';

describe('Test in Create Form component', () => {
  const setDataPhone = jest.fn();
  let wrapper = shallow(<CreatePhone />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<CreatePhone />);
  });

  test('should render component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should set input value in onChange', () => {
    const input = wrapper.find('input#phoneName');
    const value = 'iPhone 11';

    input.simulate('change', { target: { value } });
  });

  test('not should submit the form without data', () => {
    wrapper.find('form').simulate('submit', { preventDefault() {} });

    // If you want check if handleSubmit is called, remove .not, the test will fail...and it's ok because means that function have been called
    expect(setDataPhone).not.toHaveBeenCalled();
  });
});
