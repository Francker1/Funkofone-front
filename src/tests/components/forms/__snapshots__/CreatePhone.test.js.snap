// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in Create Form component should render component correctly 1`] = `
<Fragment>
  <NavBar />
  <Container
    className="list-container"
    fluid={false}
  >
    <Row
      noGutters={false}
    >
      <Col
        className="mb-4"
      >
        <h1>
          Create Phone
        </h1>
        <Link
          to="/phones"
        >
          Back to All Phones
        </Link>
      </Col>
    </Row>
    <Row
      noGutters={false}
    >
      <Col>
        <form
          onSubmit={[Function]}
        >
          <div
            className="form-row"
          >
            <div
              className="form-group col-md-6"
            >
              <label
                htmlFor="phoneName"
              >
                Phone name
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneName"
                name="name"
                onChange={[Function]}
                placeholder="Phone name"
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group col-md-6"
            >
              <label
                htmlFor="phonePrice"
              >
                Price
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phonePrice"
                name="price"
                onChange={[Function]}
                placeholder="Phone price"
                required={true}
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            className="form-row"
          >
            <div
              className="form-group col-md-4"
            >
              <label
                htmlFor="phoneManufacturer"
              >
                Manufacturer
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneManufacturer"
                name="manufacturer"
                onChange={[Function]}
                placeholder="Phone manufacturer"
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group col-md-4"
            >
              <label
                htmlFor="phoneModel"
              >
                Model
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneModel"
                name="model"
                onChange={[Function]}
                placeholder="Phone model"
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group col-md-4"
            >
              <label
                htmlFor="phoneColor"
              >
                Color
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneColor"
                name="color"
                onChange={[Function]}
                placeholder="Phone color"
                required={true}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="form-row"
          >
            <div
              className="form-group col"
            >
              <label
                htmlFor="phoneDetail"
              >
                Detail
              </label>
              <textarea
                autoComplete="off"
                className="form-control"
                id="phoneDetail"
                name="detail"
                onChange={[Function]}
                placeholder="Phone detail"
                required={true}
                value=""
              />
            </div>
          </div>
          <div
            className="form-row"
          >
            <div
              className="form-group col-md-3"
            >
              <label
                htmlFor="phoneRam"
              >
                RAM
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneRam"
                name="ram"
                onChange={[Function]}
                placeholder="RAM"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group col-md-3"
            >
              <label
                htmlFor="phoneSize"
              >
                Size
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneSize"
                name="size"
                onChange={[Function]}
                placeholder="Phone size"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group col-md-3"
            >
              <label
                htmlFor="phoneScreen"
              >
                Screen
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneScreen"
                name="screen"
                onChange={[Function]}
                placeholder="Phone screen"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group col-md-3"
            >
              <label
                htmlFor="phoneProcessor"
              >
                Processor
              </label>
              <input
                autoComplete="off"
                className="form-control"
                id="phoneProcessor"
                name="processor"
                onChange={[Function]}
                placeholder="Phone processor"
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="btn btn-primary"
            type="submit"
          >
            Create
          </button>
        </form>
      </Col>
    </Row>
  </Container>
</Fragment>
`;
